### DISCUSSION TITLE
Knowledge Distillation yolo11x to yolo11n

### DISCUSSION THREAD

**Discussion by sanjayb157** [ultralytics/ultralytics#23317] [Q&A] [answered]
Pls guide for Knowledge Distillation yolo11x to yolo11n.   from scratch.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @sanjayb157, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (knowledge distillation from YOLO11x to YOLO11n), please share as much detail as possible so we can point you in the right direction, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).✅  
1)Your teacher/student setup details (what weights you’re starting from, whether you’re training “from scratch” or fine-tuning)  
2)Your dataset type and size (detection/segmentation/etc.) and a few representative sample images  
3)Your current training command/config and any logs you already have  
4)Your hardware + OS + Python/PyTorch versions

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by sanjayb157** [reply]
1.  custom dataset, fine-tuning
2. detection
3. yolo detect train
4. os ubuntu ,  Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (NVIDIA A100-SXM4-80GB, 81222MiB) google colab

I want complete guidance for knowledge distillation from YOLO11x to YOLO11n

**Reply by glenn-jocher** [reply]
Ultralytics doesn’t currently provide an “out-of-the-box” teacher→student knowledge distillation loss for YOLO11 via `yolo train` (KD generally requires custom loss terms; see the concept overview in the [Knowledge Distillation glossary](https://www.ultralytics.com/glossary/knowledge-distillation)), so the most practical supported workflow today is pseudo-label distillation: use `yolo11x.pt` to generate YOLO-format labels on extra **unlabeled** images, then fine-tune `yolo11n.pt` on the combined labeled+pseudo-labeled dataset (baseline usage is in the [Ultralytics YOLO11 docs](https://docs.ultralytics.com/models/yolo11/)).

```python
from ultralytics import YOLO

teacher = YOLO("yolo11x.pt")
teacher.predict(source="path/to/unlabeled_images", save_txt=True, save_conf=False, conf=0.25)

student = YOLO("yolo11n.pt")
student.train(data="path/to/combined.yaml", epochs=100, imgsz=640)
```

If you confirm whether you have a separate unlabeled image pool (and share your exact `yolo detect train ...` command + class count), I can suggest sane `conf`/`iou` pseudo-label settings and a simple 2-stage training schedule.

**Reply by sanjayb157** [reply]
no seprated unlabled pool.
yolo detect train  time=22   data=data.yaml  model=yolo11s.pt multi_scale=True patience=60  epochs=200 batch=128 imgsz=640     cache=False       
my class count is 10.

**Reply by glenn-jocher** [reply]
If you don’t have any additional unlabeled images, there’s effectively nothing to “distill” beyond standard fine-tuning (true teacher→student KD needs custom loss terms; the supported practical alternative is pseudo-label/self-training, which requires extra unlabeled data—see the concept in this [semi-supervised learning overview](https://www.ultralytics.com/glossary/semi-supervised-learning)). In your case, the best supported path is to train the smaller model directly on your labeled set (swap to `yolo11n.pt` if you want nano):

```bash
yolo detect train data=data.yaml model=yolo11n.pt epochs=200 batch=128 imgsz=640 patience=60 multi_scale=True cache=False
```

If you can later export extra frames/images from the same domain (even from the same videos), share roughly how many you can collect and I can suggest conservative pseudo-label `conf` settings for `yolo11x.pt` before training `yolo11n.pt` on the merged set.

**Reply by sanjayb157** [reply]
10% images seprated from dataset. And removed txt files.
pls guide for next  10 classes.

**Reply by glenn-jocher** [reply, marked as answer]
If that “unlabeled” 10% is a separate image folder with the `.txt` removed, you can do pseudo-label distillation by (1) training a strong **teacher** on the remaining 90% labeled data, then (2) running `predict` with `save_txt=True` to generate YOLO labels for those images (same 10 classes / same `names` order as your `data.yaml`), then (3) training `yolo11n.pt` on the combined set; see the [Predict mode docs](https://docs.ultralytics.com/modes/predict/) for output locations and args.

```bash
yolo detect train data=data.yaml model=yolo11x.pt epochs=200 imgsz=640 batch=128 patience=60 multi_scale=True cache=False
```

```bash
yolo detect predict model=runs/detect/train/weights/best.pt source=/path/to/unlabeled_images save_txt=True conf=0.6 iou=0.7
```

Copy the generated label files from `runs/detect/predict*/labels/` into your dataset’s corresponding `labels/` folder next to those 10% images, delete any `labels/*.cache`, then train the student normally (details in the [Train mode docs](https://docs.ultralytics.com/modes/train/)).

**Reply by sanjayb157** [reply]
Thank you

**Reply by glenn-jocher** [reply]
You’re welcome — if anything looks off in the pseudo-labels (missing boxes or wrong class IDs), post your exact `yolo detect predict ...` command, `data.yaml`, and `!yolo checks` output here and we can help debug; the [YOLO Common Issues guide](https://docs.ultralytics.com/guides/yolo-common-issues/) is also a good quick sanity check.
